File: DefaultMarginDisabler/DefaultMarginDisabler.m
//
//  DefaultMarginDisabler.m
//  DefaultMarginDisabler
//
//  Created by abc on 8/21/16.
//  Copyright © 2016 Manabu Nakazawa. All rights reserved.
//

#import "DefaultMarginDisabler.h"

static DefaultMarginDisabler *sharedPlugin;

@implementation DefaultMarginDisabler

#pragma mark - Initialization

+ (void)pluginDidLoad:(NSBundle *)plugin
{
    sharedPlugin = [[self alloc] initWithBundle:plugin];
}

+ (instancetype)sharedPlugin
{
    return sharedPlugin;
}

- (id)initWithBundle:(NSBundle *)bundle
{
    if (self = [super init])
    {
        _bundle = bundle;

        if (NSApp && !NSApp.mainMenu)
        {
            [[NSNotificationCenter defaultCenter] addObserver:self
                                                     selector:@selector(applicationDidFinishLaunching:)
                                                         name:NSApplicationDidFinishLaunchingNotification
                                                       object:nil];
        }
        else
        {
            [self initializeAndLog];
        }
    }
    return self;
}

- (void)applicationDidFinishLaunching:(NSNotification *)notification
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:NSApplicationDidFinishLaunchingNotification object:nil];
    [self initializeAndLog];
}

- (void)initializeAndLog
{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(popoverWillShowNotificationListener:) name:NSPopoverWillShowNotification object:nil];
}

#pragma mark - Implementation

- (void)popoverWillShowNotificationListener:(NSNotification *)notification
{
    NSPopover *popover = notification.object;

    if (![popover isKindOfClass:NSPopover.class])
    {
        return;
    }

    NSArray<NSView *> *views = popover.contentViewController.view.subviews.firstObject.subviews;

    if (!views)
    {
        return;
    }

    NSButton *btn = (NSButton *)[views filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(NSView *evaluatedObject, NSDictionary<NSString *, id> *bindings)
    {
        return [evaluatedObject isKindOfClass:[NSButton class]]
            && [[(NSButton *)evaluatedObject title] isEqualToString:@"Constrain to margins"];
    }]].firstObject;

    if (!btn || !btn.enabled || btn.state != NSOnState)
    {
        return;
    }

    btn.state = NSOffState;

    #pragma clang diagnostic push
    #pragma clang diagnostic ignored "-Warc-performSelector-leaks"

    [btn.target performSelector:btn.action];

    #pragma clang diagnostic pop
}

@end

File: DefaultMarginDisabler/DefaultMarginDisabler.swift
//
//  DefaultMarginDisabler.swift
//  DefaultMarginDisabler
//
//  Created by mshibanami on 2015/11/19.
//  Copyright © 2015 mshibanami. All rights reserved.
//

import Foundation
import Cocoa

class DefaultMarginDisabler: NSObject {
  private var bundle: NSBundle

  init(bundle: NSBundle) {
    self.bundle = bundle
    super.init()

    NSNotificationCenter.defaultCenter().addObserver(self, selector: "xcodeDidFinishLaunching:", name: NSApplicationDidFinishLaunchingNotification, object: nil)
  }

  deinit {
    NSNotificationCenter.defaultCenter().removeObserver(self)
  }

  func xcodeDidFinishLaunching(notification: NSNotification) {
    let center = NSNotificationCenter.defaultCenter()

    center.removeObserver(self, name: NSApplicationDidFinishLaunchingNotification, object: nil)
    center.addObserver(self, selector: "popoverWillShowNotificationListener:", name: NSPopoverWillShowNotification, object: nil)
  }

  func popoverWillShowNotificationListener(notification: NSNotification) {

    if let popover = notification.object as? NSPopover,
       let views = popover.contentViewController?.view.subviews.first?.subviews {

      let buttons = views.filter {
        $0.isKindOfClass(NSButton) && ($0 as! NSButton).title == "Constrain to margins"
      } as! [NSButton]

      if let btn = buttons.first,
         let t = btn.target {

        guard btn.enabled && btn.state == NSOnState else {
          return
        }

        btn.state = NSOffState
        t.performSelector(btn.action)
      }
    }
  }
}
